ports/unix/main.c

```diff
@@ -297,18 +234,25 @@ static int do_repl(void) {
             line = line3;
         }
 
-        int ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
+        ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
         free(line);
-        if (ret & FORCED_EXIT) {
-            return ret;
-        }
     }
-
     #endif
+    return ret;
 }
 
+
 static int do_file(const char *file) {
-    return execute_from_lexer(LEX_SRC_FILENAME, file, MP_PARSE_FILE_INPUT, false);
+    int ret = pyexec_file(file);
+    // pyexec returns 1 for success, 0 for exception, PYEXEC_FORCED_EXIT for SystemExit
+    // Convert to unix port's expected codes: 0 for success, 1 for exception, FORCED_EXIT|val for SystemExit
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret == PYEXEC_FORCED_EXIT) {
+        return FORCED_EXIT; // SystemExit (without exit value for now)
```

Comment:
The value of `SystemExit` must be propagated out of this function, otherwise `sys.exit(1)` doesn't return an error code to the shell.